### DISCUSSION TITLE
How to avoid blurry SVGs?

### DISCUSSION THREAD

**Discussion by Dremora** [fabricjs/fabric.js#9424] [Q&A] [answered]
In the example below I'm rendering the same figure, first using SVG parsed from a string, and then using manually constructed `Path` object. The former appears blurry, the latter doesn't.

https://stackblitz.com/edit/vitejs-vite-cwfnjd?file=src%2Ffabric.ts&terminal=dev

<img width="188" alt="CleanShot 2023-10-19 at 16 56 18@2x" src="https://github.com/fabricjs/fabric.js/assets/242202/40841cc1-97b9-48d4-8819-b27e8cbfab16"/>

What should I do to avoid blurry SVG?

(I'm using Fabric.js v6).

**Update**: by serialising canvas to JSON, I've figured out the difference. SVG has `top` and `left` properties set to 0.75. Manually setting them to 0 resolves the issue. But where does 0.75 come from?

**Comment by ShaMan123** [comment]
Good catch, rounding will do the difference
You can use `console.trace()` - but it comes from the svg parser

**Reply by Dremora** [reply]
Unfortunately, with more complex SVGs, rounding doesn't help anymore.

But there should be a way to resolve it, at least in theory. We use Figma for design, and no matter the zoom level, all vector graphics always appears sharp there.

**Reply by asturur** [reply, marked as answer]
so the issue is that the path gets cached as crisp as possible, but then if it is place across pixels it will be a bit blurry.
You can disable caching or you can place it at top left as integers.
Is a tradeoff between performance and crispness, or as you mentioned you can get the Figma design engine and build on top of that that i think is way more powerful and complex than fabricjs

**Reply by asturur** [reply]
The 0.75 comes from how the path is defined.
```
m96.3.752 6.188 58.366a19.998 19.998 0 0 0 12.883 16.625L170.346 96.3l-58.366 6.187a19.998 19.998 0 0 0-16.625 12.883l-20.558 54.975-6.187-58.366a20 20 0 0 0-12.883-16.625L.752 74.797l58.366-6.187a20 20 0 0 0 16.625-12.883L96.3.752Z
```
the svg path doesn't have a x and y property to place a path, the top and left final position is determined by how all its commands are placed on the canvas.

**Reply by Dremora** [reply]
Thanks! Disabling caching seems to have resolved the issue. Performance is not much of a concern, since we won't have many elements.

In regards to Figma design engine, I don't think it's open-source. Of course, given Figma's budget, it's no surprise that they have managed to build something that combines scale, performance and rendering quality without any visible tradeoffs. Also, as far as I know, it's written in C++ and is compiled to WebAssembly.

**Reply by asturur** [reply]
No is not open source, but probably with an amount of money can be bought/licensed?
i wanted to stress out we were comparing things at a different scale entirely.

**Reply by Dremora** [reply]
Yes, I understand that fabric.js is a free open-source library, and I'm grateful we have it.

**Reply by ShaMan123** [reply]
Non the less I am sure we can come up with solutions to these problems, not now as we are invested in other stuff, but contributors can experiment
